flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCheckpointIDCounterITCase.java

```diff
@@ -66,6 +70,56 @@ public void testShutdownRemovesState() throws Exception {
         assertThat(client.checkExists().forPath(counter.getPath())).isNull();
     }
 
+    @Test
+    public void testShutdownForLocallyTerminatedJobStatus() throws Exception {
+        ZooKeeperCheckpointIDCounter counter = createCheckpointIdCounter();
+        counter.start();
+
+        CuratorFramework client = zookeeper.getClient();
+        assertThat(client.checkExists().forPath(counter.getPath())).isNotNull();
+
+        counter.shutdown(JobStatus.SUSPENDED).join();
+        assertThat(client.checkExists().forPath(counter.getPath())).isNotNull();
+    }
+
+    @Test
+    public void testIdempotentShutdown() throws Exception {
+        ZooKeeperCheckpointIDCounter counter = createCheckpointIdCounter();
+        counter.start();
+
+        CuratorFramework client = zookeeper.getClient();
+        client.delete().forPath(counter.getPath());
+
+        // shutdown shouldn't fail due to missing path
+        counter.shutdown(JobStatus.FINISHED).join();
+        assertThat(client.checkExists().forPath(counter.getPath())).isNull();
+    }
+
+    @Test
+    public void testShutdownWithFailureDueToMissingConnection() throws Exception {
+        ZooKeeperCheckpointIDCounter counter = createCheckpointIdCounter();
+        counter.start();
+
+        cleanAndStopZooKeeperIfRunning();
+
+        assertThat(counter.shutdown(JobStatus.FINISHED)).failsWithin(Duration.ofHours(1));
+    }
+
+    @Test
+    public void testShutdownWithFailureDueToChildNodes() throws Exception {
```

Comment:
DueTo**Missing**ChildNode

Reply:
No, the deletion is failing because there are child nodes which I added for the test case (and I removed the recursive deletion). The IDCounter shouldn't have child nodes in general...

Reply:
DueTo**Existing**ChildNode then.

Why does the comment say "// shutdown shouldn't fail due to missing path"?
What is actual ZK behavior? Will it happily delete the unexpected child node as well?